### ISSUE TITLE
TaskOrchestrationContext DataConverters settings customization

### ISSUE THREAD

**Issue by known-as-bmf** [Azure/durabletask#390]
# TL;DR
Would it be possible to enable the customization of `JsonSerializerSettings` for what goes in and out of orchestrators ?

# Intro
In [Azure Function Durable Extension](https://github.com/Azure/azure-functions-durable-extension), it is possible to customize the serialization/deserialization of what goes in and out of an activity/entity(?) by overloading `IMessageSerializerSettingsFactory` and `IErrorSerializerSettingsFactory` on the DI container.

However, it looks like it is not possible to customize how an orchestrator serializes the input of an activity. It looks like it's hardcoded here:

- https://github.com/Azure/durabletask/blob/master/src/DurableTask.Core/TaskOrchestrationContext.cs#L51
- https://github.com/Azure/durabletask/blob/master/src/DurableTask.Core/Serializing/JsonDataConverter.cs#L32

Here is a schema to illustrate what it looks like:
- In green: serialization/deserialization we have "control" over
- In red: serialization/deserialization we do not have "control" over

![diagram](https://user-images.githubusercontent.com/11721166/79616957-9f435b80-8106-11ea-8bd7-a0a3d514356e.png)

Would it be possible to enable the customization of `JsonSerializerSettings` for what goes in and out of orchestrators ?

# Why ?

## use case

1. an activity is producing a result (an object, list, dict...)
2. a *saving* activity takes this result as input (here the input is typed as `object`), serializes it to json and saves it in an azure blob.

The activity in step 2 will be used in many scenarios and with many different objects as input (hence the `object` typing).

## problem

1. The orchestrator will serialize the input of the 1st activity (adding `$type` properties due to `TypeNameHandling.Objects` rule [here](https://github.com/Azure/durabletask/blob/master/src/DurableTask.Core/Serializing/JsonDataConverter.cs#L35)) and pass that to the 2nd activity.
2. The 2nd activity will deserializez this to `object` (in reallity, to a JToken) INCLUDING those `$type`  properties.
3. This is serialized another time before being inserted in the blob
4. The `$type` properties will end up in the blob file.

**Comment by ConnorMcMahon** [comment]
We actually started utilizing these in [this PR](https://github.com/Azure/azure-functions-durable-extension/pull/1294/files), which is in the v2.2.1 release.

**Comment by known-as-bmf** [comment]
You are absolutely right, sorry!

What got me is: 
- if you **do** overload  `IMessageSerializerSettingsFactory`, then the settings are shared between durable function extension and durable task.
- if you do **not** overload `IMessageSerializerSettingsFactory`, [durable function extension](https://github.com/Azure/azure-functions-durable-extension/blob/dev/src/WebJobs.Extensions.DurableTask/MessageSerializerSettingsFactory.cs) and [durable task](https://github.com/Azure/durabletask/blob/master/src/DurableTask.Core/Serializing/JsonDataConverter.cs) dont share the same defaults !

And I did **not** overload it!

Now to explain why I didn't actually overload `IMessageSerializerSettingsFactory`: the durable function extension defaults were what I wanted to use 😅

I naively thougt that durable function extension would propagate its defaults regardless of the [`IsDefault` flag](https://github.com/Azure/azure-functions-durable-extension/blob/dev/src/WebJobs.Extensions.DurableTask/Listener/TaskEntityShim.cs#L157). Maybe that should be the case ?

Anyway, thanks for your help!

**Comment by ConnorMcMahon** [comment]
Ah yes, that is a tricky sticking point. We were debating about changing the defaults of Durable Function extension to match DurableTask, but we are always a bit hesitant about changing our serializer settings in a non-breaking release, as it could theoretically break some customer's inflight orchestrations. We decided against fixing this discrepancy for this reason, and just made sure we use any settings from our message/error serializer settings factory in both places.

If we can prove that we can safely use the same settings in both locations without breaking any customers, I would be very happy to fix that discrepency.

As a note, since this topic is a bit more Durable Functions than Durable Task framework, I am going to move this issue to that repo.
